File: server/schemas/product.py
from pydantic import BaseModel


class ProductVariantOut(BaseModel):
    id: str
    size: str | None = None
    shipping_mode: str | None = None
    sale_price: float | None = None
    compare_at_price: float | None = None


class ProductImageOut(BaseModel):
    url: str
    name: str | None = None
    alt: str | None = None
    sort_order: int | None = None


class ProductSearchResult(BaseModel):
    id: str
    name: str
    brand: str | None = None
    product_type: str | None = None
    price: float | None = None
    image_url: str | None = None


class ProductOut(BaseModel):
    id: str
    name: str
    brand: str | None = None
    product_type: str | None = None
    product_category: str | None = None
    sku: str | None = None
    price: float | None = None
    compare_at_price: float | None = None
    currency: str = "INR"
    description: str | None = None
    images: list[ProductImageOut] = []
    variants: list[ProductVariantOut] = []

File: client/src/types/product.ts
export type ProductSearchResult = {
  id: string;
  name: string;
  brand: string | null;
  product_type: string | null;
  price: number | null;
  image_url: string | null;
};

export type ProductVariant = {
  id: string;
  size: string | null;
  shipping_mode: string | null;
  sale_price: number | null;
  compare_at_price: number | null;
};

export type ProductImage = {
  url: string;
  name: string | null;
  alt: string | null;
  sort_order: number | null;
};

export type Product = {
  id: string;
  name: string;
  brand: string | null;
  product_type: string | null;
  product_category: string | null;
  sku: string | null;
  price: number | null;
  compare_at_price: number | null;
  currency: string;
  description: string | null;
  images: ProductImage[];
  variants: ProductVariant[];
};
